scripts/Makefile.build

```diff
@@ -333,23 +333,53 @@ rust_cross_flags := --target=$(realpath $(KBUILD_RUST_TARGET))
 
 rust_allowed_features := allocator_api,bench_black_box,concat_idents,generic_associated_types,global_asm
 
+rust_common_cmd = \
+	RUST_MODFILE=$(modfile) $(RUSTC_OR_CLIPPY) \
+	$(rust_flags) $(rust_cross_flags) \
+	-Zallow-features=$(rust_allowed_features) \
+	-Zcrate-attr=no_std \
+	-Zcrate-attr='feature($(rust_allowed_features))' \
+	--extern alloc --extern kernel \
+	--crate-type rlib --out-dir $(obj) -L $(objtree)/rust/ \
+	--crate-name $(basename $(notdir $@))
+
+rust_handle_depfile = \
+	mv $(obj)/$(basename $(notdir $@)).d $(depfile); \
+	sed -i '/^\#/d' $(depfile)
+
 quiet_cmd_rustc_o_rs = $(RUSTC_OR_CLIPPY_QUIET) $(quiet_modtag) $@
       cmd_rustc_o_rs = \
-	RUST_MODFILE=$(modfile) \
-	$(RUSTC_OR_CLIPPY) $(rust_flags) $(rust_cross_flags) \
-		--emit=dep-info,obj,metadata \
-		-Zallow-features=$(rust_allowed_features) \
-		-Zcrate-attr=no_std \
-		-Zcrate-attr='feature($(rust_allowed_features))' \
-		--extern alloc --extern kernel \
-		--crate-type rlib --out-dir $(obj) -L $(objtree)/rust/ \
-		--crate-name $(patsubst %.o,%,$(notdir $@)) $<; \
-	mv $(obj)/$(subst .o,,$(notdir $@)).d $(depfile); \
-	sed -i '/^\#/d' $(depfile)
+	$(rust_common_cmd) --emit=dep-info,obj,metadata $<; \
+	$(rust_handle_depfile)
 
 $(obj)/%.o: $(src)/%.rs FORCE
 	$(call if_changed_dep,rustc_o_rs)
 
+quiet_cmd_rustc_i_rs = $(RUSTC_OR_CLIPPY_QUIET) $(quiet_modtag) $@
+      cmd_rustc_i_rs = \
+	$(rust_common_cmd) --emit=dep-info -Zunpretty=expanded $< >$@; \
+	command -v $(RUSTFMT) >/dev/null && $(RUSTFMT) $@; \
+	$(rust_handle_depfile)
+
+$(obj)/%.i: $(src)/%.rs FORCE
+	$(call if_changed_dep,rustc_i_rs)
+
+quiet_cmd_rustc_s_rs = $(RUSTC_OR_CLIPPY_QUIET) $(quiet_modtag) $@
+      cmd_rustc_s_rs = \
+	$(rust_common_cmd) --emit=dep-info,asm $<; \
```

Comment:
One thing to note is that `--emit asm` and `--emit llvm-ir` force a single codegen unit for backwards compatibility reasons. This can affect optimizations.

Reply:
Thanks! That is good to know. Currently, we use `-Ccodegen-units=1`, so it should be the same. But that may not always be the case, so I think it is worth noting. (Is there some issue/link on Rust about this to follow/track/include?).

Having said that, if it is just a backwards compatibility reason, why does not `rustc` provide a new/fixed `asm`/`llvm-ir`?

Reply:
> (Is there some issue/link on Rust about this to follow/track/include?).

Not that I know of.

> Having said that, if it is just a backwards compatibility reason, why does not rustc provide a new/fixed asm/llvm-ir?

No particular reason I think except maybe that `--emit` currently doesn't have a way for a single emit type to create multiple files. I guess it would need something like `--emit llvm-ir=my_crate.%.ll` where `%` would be replaced with a different value for each cgu. This reminds me that the current usage of `--emit obj` also forces a single codegen unit, so it is consistent with `--emit asm` and `--emit llvm-ir`. All of them force a single codegen unit for the same reason so any method to allow multiple codegen units can be applied to all of them without risking any of them getting out of sync with regards to cgu count.

Reply:
> No particular reason I think except maybe that `--emit` currently doesn't have a way for a single emit type to create multiple files.

So, if I understand correctly, the backwards compatibility issue you mentioned is that `--emit` will always have to imply `-Ccodegen-units=1` since it already does now (though giving `-Ccodegen-units` overrides that, giving a warning -- I recall I added `-Ccodegen-units=1` back then precisely due to the diagnostic).

> I guess it would need something like `--emit llvm-ir=my_crate.%.ll` where `%` would be replaced with a different value for each cgu.

Makes a lot of sense. Given `--emit` is cumulative, one could give a different one per kind of output. Otherwise it could accept something like `--emit=llvm-ir=my_crate.%.ll,asm=my_crate.%.s` (I recall seeing a discussion about this kind of "multiple `=`/key-values" for something, not sure for which flag it was -- though it is not particularly pretty).

> This reminds me that the current usage of `--emit obj` also forces a single codegen unit, so it is consistent with `--emit asm` and `--emit llvm-ir`. All of them force a single codegen unit for the same reason so any method to allow multiple codegen units can be applied to all of them without risking any of them getting out of sync with regards to cgu count.

I have added a note nearby the single targets to note this discussion -- please take a look!